### DISCUSSION TITLE
API based k3s management

### DISCUSSION THREAD

**Discussion by xavier83** [k3s-io/k3s#8574] [General]
**Is your feature request related to a problem? Please describe.**

Currently, management of k3s installation and maintenance is only possible on a local instance shell and it is static. This make "re-shaping" the cluster node configurations a tedious task, effectively making each k3s node a _pet_ rather than _cattle_.

**Describe the solution you'd like**

 It would be incredibly useful and powerful if k3s agent/worker role and configurations could be managed remotely. This could be enabled  by implementing a subset of [talos api](https://github.com/siderolabs/talos/tree/main/api) such as upgrading, bootstrapping, resetting, discovery? etc and a management tool similar to [talosctl](https://www.talos.dev/v1.5/learn-more/talosctl/). It could potentially allow k3s clusters to resize by adding nodes incrementally as needed similar to cluster autoscaler.

**Describe alternatives you've considered**

[talos linux](https://www.talos.dev/) requires going _all in_ including the host os itself which isn't feasible always.

**Additional context**

This inspired by [talos linux](https://www.talos.dev/)

**Comment by brandond** [comment]
This sort of stuff is all covered by Cluster API (or a competing standard like the talos API), and is not usually part of the distro itself. Rancher handles CAPI for K3s and RKE2, although it is not currently easy to use directly. There is a 3rd party standalone implementation under development at https://github.com/cluster-api-provider-k3s/cluster-api-k3s
